packages/block-editor/src/hooks/auto-inserting-blocks.js

```diff
@@ -0,0 +1,214 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { addFilter } from '@wordpress/hooks';
+import { Fragment } from '@wordpress/element';
+import { PanelBody, ToggleControl } from '@wordpress/components';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { createBlock, store as blocksStore } from '@wordpress/blocks';
+import { useDispatch, useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { InspectorControls } from '../components';
+import { store as blockEditorStore } from '../store';
+
+function AutoInsertingBlocksControl( props ) {
+	const autoInsertedBlocksForCurrentBlock = useSelect(
+		( select ) => {
+			const { getBlockTypes } = select( blocksStore );
+			return getBlockTypes()?.filter(
+				( block ) =>
+					block.autoInsert &&
+					Object.keys( block.autoInsert ).includes( props.blockName )
+			);
+		},
+		[ props.blockName ]
+	);
+
+	const {
+		autoInsertedBlockClientIds,
+		blockIndex,
+		rootClientId,
+		innerBlocksLength,
+	} = useSelect(
+		( select ) => {
+			const { getBlock, getBlockIndex, getBlockRootClientId } =
+				select( blockEditorStore );
+			const _rootClientId = getBlockRootClientId( props.clientId );
+
+			const _autoInsertedBlockClientIds =
+				autoInsertedBlocksForCurrentBlock.reduce( ( acc, block ) => {
+					const relativePosition =
+						block?.autoInsert?.[ props.blockName ];
+					let autoInsertedBlock;
+					if ( [ 'before', 'after' ].includes( relativePosition ) ) {
+						const siblings = getBlock( _rootClientId )?.innerBlocks;
+
+						// Any of the current block's siblings (with the right block type) qualifies
+						// as an auto-inserted block (inserted `before` or `after` the current one),
+						// as the block might've been auto-inserted and then moved around a bit by the user.
+						autoInsertedBlock = siblings.find(
+							( { name } ) => name === block.name
+						);
+					} else if (
+						[ 'first_child', 'last_child' ].includes(
+							relativePosition
+						)
+					) {
+						const { innerBlocks } = getBlock( props.clientId );
+
+						// Any of the current block's child blocks (with the right block type) qualifies
+						// as an auto-inserted first or last child block, as the block might've been
+						// auto-inserted and then moved around a bit by the user.
+						autoInsertedBlock = innerBlocks.find(
+							( { name } ) => name === block.name
+						);
+					}
```

Comment:
another note on these if-array-check-condition blocks. thinking of types. did you consider using a `switch` on the legitimate values of the relative position?

```js
switch ( relativePosition ) {
	case 'before':
	case 'after':
		// I am a comment blah blah blah.
		autoInsertedBlock = getBlock( _rootClientId )?.innerBlocks )
			.find( ( { name } ) => name === block.name );
		break;

	case 'first_child':
	case 'last_child':
		autoInsertedBlock = getBlock( props.clientId ).innerBlock )
			.find( ( { name } ) => name === block.name );
		break;
}
```

in this case if we add an `as const` type then I think TypeScript should alert us if we overlook one of the values.

then again, written this way, it looks like both halves here have more symmetry than we expose. makes me wonder.

```js
const containingBlockClientId = ( 'before' === relativePosition || 'after' === relativePosition )
	? _rootClientId
	: props.clientId;

const autoInsertedBlock = getBlock( containingBlockClientId )
	.innerBlocks
	.find( ( { name } ) => name === block.name );
```

Reply:
That symmetry did occur to me when writing this, but I wasn't sure if rewriting it that way would make it that much more legible; and what's more, I was somewhat concerned it would put the focus too much on something that's actually somewhat incidental: We just happen to offer insertion in those relative positions; but in principle it's possible that we'd offer different ones (e.g. nth child, or in-between-blocks). While I don't really expect something like that to be added, I didn't want to put too much emphasis on this incidental symmetry, as it might suggest to people reading and modifying the code that they have to follow a certain pattern for different insertion positions.

I agree that duplicating the `.find()` stuff is a bit annoying; I'll let this simmer for a bit and maybe try on different variations for size.

Reply:
Here's a bit of a hybrid based on the two suggestions you made that looked kinda good to me: 7dd82006fa